azure-storage-cpp-lite/src/blob/blob_client_wrapper.cpp

```diff
@@ -623,108 +630,101 @@ namespace microsoft_azure {
+                    r = pwrite(fd, os.str().c_str(), firstChunk.response().size, 0);
+                }
+                if (-1 == r) {
+                    close(fd);
+                    return;
+                }
+                close(fd);
 
+                // Download the rest.
+                const auto left = length - firstChunk.response().size;
+                const auto chunk_size = std::max(DOWNLOAD_CHUNK_SIZE, (left + downloaders - 1)/ downloaders);
+                std::vector<std::future<int>> task_list;
+                for(unsigned long long offset = firstChunk.response().size; offset < length; offset += chunk_size)
+                {
+                    const auto range = std::min(chunk_size, length - offset);
+                    auto single_download = std::async(std::launch::async, [originalEtag, offset, range, this, &destPath, &container, &blob](){
+                        // Note, keep std::ios_base::in to prevent truncating of the file.
+                        auto output = std::ofstream(destPath.c_str(), std::ios_base::out |  std::ios_base::in);
+                        output.seekp(offset);
+                        auto chunk = m_blobClient->get_chunk_to_stream_sync(container, blob, offset, range, output);
+                        output.close();
+                        if(!chunk.success())
                         {
-                            std::lock_guard<std::mutex> lock(mutex);
-                            ++parallel;
-                            //std::cout << "parallel done: " << parallel << std::endl;
-                            cv.notify_one();
+                           // Looks like the blob has been replaced by smaller one - ask user to retry.
+                           if (constants::code_request_range_not_satisfiable == chunk.error().code) {
+                              return EAGAIN;
+                           }
+                           return std::stoi(chunk.error().code);
+                        }
+                        // The etag has been changed - ask user to retry.
+                        if (originalEtag != chunk.response().etag) {
+                           return EAGAIN;
                         }
-                        return errno;
+                        // Check for any writing errors.
+                        if (!output) {
+                            return unknown_error;
+                        }
+                        return 0;
                     });
                     task_list.push_back(std::move(single_download));
                 }
 
+                // Wait for workers to complete downloading.
                 for(size_t i = 0; i < task_list.size(); ++i)
                 {
                     task_list[i].wait();
+                    auto result = task_list[i].get();
+                    // let's report the first encountered error for consistency.
+                    if (0 != result && errcode == 0) {
+                        errcode = result;
```

Comment:
If error happens in downloading, no need to wait for all tasks to finish.

Reply:
not sure. we cannot just return an error leaving threads working and there is no cancellation mechanism available. So, the best thing i can do is wait.